ports/cheriot-rtos/hal/uart.cpp

```diff
@@ -0,0 +1,42 @@
+#include <platform-uart.hh>
+
+#include "timer.h"
+
+extern "C" uint8_t uart_get_rx_level(volatile OpenTitanUart<115'200> *block) {
+    return (block->fifoStatus >> 16) & 0xff;
+}
+
+extern "C" uint8_t uart_get_tx_level(volatile OpenTitanUart<115'200> *block) {
+    return block->fifoStatus & 0xff;
+}
+
+extern "C" bool uart_is_readable(volatile OpenTitanUart<115'200> *block) {
+    return block->can_read();
+}
+
+extern "C" bool uart_is_writable(volatile OpenTitanUart<115'200> *block) {
+    return block->can_write();
+}
+
+extern "C" bool uart_read(volatile OpenTitanUart<115'200> *block, uint8_t *out, uint32_t timeout_ms) {
+    uint32_t t_end = get_time() + MS_TO_CLOCK_CYCLES(timeout_ms);
+
+    while (!block->can_read()) {
+        if (get_time() > t_end) {
+            return false;
+        }
+    }
+
+    *out = block->rData;
+    return true;
+}
```

Comment:
It may be worth asking if CHERIoT RTOS is willing to accept a PR for non-blocking (or timeout-based) reads to clean this up.